interp/interp.go

```diff
@@ -0,0 +1,158 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package interp
+
+import (
+	"sort"
+)
+
+// Predictor predicts the value of a function. It handles both
+// interpolation and extrapolation.
+type Predictor interface {
+	// Predict returns the predicted value at x.
+	Predict(float64) float64
+}
+
+// Fitter fits a predictor to data.
+type Fitter interface {
+	// Fit fits a predictor to (X, Y) value pairs provided as two slices.
+	// It panics if len(xs) < 2, elements of xs are not strictly increasing or len(xs) != len(ys).
+	Fit(xs, ys []float64)
+}
+
+// FittablePredictor is a Predictor which can fit itself to data.
+type FittablePredictor interface {
+	Fitter
+	Predictor
+}
+
+// Constant predicts a constant value.
+type Constant struct {
+	// Constant Y value.
+	Value float64
+}
+
+// Predict returns the predicted value at x.
+func (c Constant) Predict(x float64) float64 {
+	return c.Value
+}
+
+// PiecewiseLinear is a piecewise linear 1-dimensional interpolator.
+// It extrapolates flat forwards (backwards) the last (first) known Y value.
+type PiecewiseLinear struct {
+	// Interpolated X values.
+	xs []float64
+
+	// Interpolated Y data values, same len as ys.
+	ys []float64
+
+	// Slopes of Y between neighbouring X values. len(slopes) + 1 == len(xs) == len(ys).
+	slopes []float64
+}
+
+// Fit fits a piecewise linear predictor to (X, Y) value pairs provided as two slices.
+// It panics if len(xs) < 2, elements of xs are not strictly increasing or len(xs) != len(ys).
+func (pl *PiecewiseLinear) Fit(xs, ys []float64) {
+	n := len(xs)
+	if len(ys) != n {
+		panic("interp: xs and ys have different lengths")
+	}
+	if n < 2 {
+		panic("interp: too few points for interpolation")
+	}
+	m := n - 1
+	pl.slopes = make([]float64, m)
+	for i := 0; i < m; i++ {
+		dx := xs[i+1] - xs[i]
+		if dx <= 0 {
+			panic("interp: xs values not strictly increasing")
+		}
+		pl.slopes[i] = (ys[i+1] - ys[i]) / dx
+	}
+	pl.xs = xs
+	pl.ys = ys
+}
+
+// Predict returns the interpolation value at x.
+func (pl PiecewiseLinear) Predict(x float64) float64 {
+	i := findSegment(pl.xs, x)
+	if i < 0 {
+		return pl.ys[0]
+	}
+	// i < len(pci.xs)
+	xI := pl.xs[i]
+	if x == xI {
+		return pl.ys[i]
+	}
+	n := len(pl.xs)
+	if i == n-1 {
+		// x > li.xs[i]
+		return pl.ys[n-1]
+	}
+	// i < len(i1d.xs) - 1
+	return pl.ys[i] + pl.slopes[i]*(x-xI)
+}
+
+// PiecewiseConstant is a piecewise constant 1-dimensional interpolator.
+// It extrapolates flat forwards (backwards) the last (first) known Y value.
+type PiecewiseConstant struct {
+	// Whether the interpolated function is left- or right-continuous.
+	// If LeftContinuous == true, then y(xs[i]) == y(xs[i] - eps) for small eps > 0. Otherwise,
+	// y(xs[i]) == y(xs[i] + eps).
+	LeftContinuous bool
```

Comment:
If you want to model stochastic processes, their trajectories can be either left- or right-continuous. We can provide a default value for `LeftContinuous`, which should probably be `false`.

Reply:
I'm still not clear why it needs to be alterable. If this is defined as an unexported field set at construction, the details don't need to be propagated to other types in a visible way.

Reply:
Initially it was not alterable, but then one of the reviewers requested to export it... I can hide it back.

Reply:
Yes, Brendan did. If there is a reason for it to be exported, then fine, but it hasn't been given yet.

Reply:
I'll let you two come to a consensus and do what you ask ;-)

Reply:
In floating-point arithmetic, does it matter if the function is left or right continuous? If not, it could be omitted even from construction.

Reply:
It does. There are many numbers which can be represented exactly in floating point. For example, 0.

Reply:
The reason why I suggested in be exported was to do as we do in the stat/sample package, something like

```
f := PiecewiseConstant{LeftContinuous: True}.Fit(x, y)
```

That said, I think I'm with Vladimir. I understand Roman's point that it can matter, but I think having it be left continuous is far more common, especially for novice users. You want it to be the typical case that if I give an {x, y} point, the prediction at x is y. 

It seems to me that the best thing to do for this package is to leave it out. For a specialized package like stochastic processes, it is easy to add an independent `RightContiguous` type that fits the interface. If we find it is sufficiently common to need, we could possibly add it here. It seems this is more complicated than a standard user would need though.